identity-core/src/main/java/com/auth0/identity/AuthorizedIdentityProvider.java

```diff
@@ -0,0 +1,146 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.auth0.identity;
+
+import android.app.Activity;
+import android.content.Intent;
+import android.support.annotation.NonNull;
+
+import com.auth0.core.Application;
+import com.auth0.identity.util.PermissionCallback;
+import com.auth0.identity.util.PermissionHandler;
+
+import java.util.List;
+
+/**
+ * Interface for allowing an IdentityProvider to request specific permissions
+ * before calling start.
+ */
+public abstract class AuthorizedIdentityProvider implements IdentityProvider, PermissionCallback {
+
+    protected IdentityProvider identityProvider;
+    private Activity activity;
+    private String serviceName;
+    private PermissionHandler handler;
+
+    public AuthorizedIdentityProvider(@NonNull IdentityProvider provider) {
+        this.identityProvider = provider;
+    }
+
+    @Override
+    public void setCallback(IdentityProviderCallback callback) {
+        identityProvider.setCallback(callback);
+    }
+
+    @Override
+    public void start(Activity activity, IdentityProviderRequest request, Application application) {
+        checkPermissions(activity, request.getServiceName(), true);
+    }
+
+    @Override
+    public void start(Activity activity, @NonNull String connectionName) {
+        checkPermissions(activity, connectionName, true);
+    }
+
+    @Override
+    public void stop() {
+        identityProvider.stop();
+    }
+
+    @Override
+    public boolean authorize(Activity activity, int requestCode, int resultCode, Intent data) {
+        return identityProvider.authorize(activity, requestCode, resultCode, data);
+    }
+
+    @Override
+    public void clearSession() {
+        identityProvider.clearSession();
+    }
+
+    /**
+     * Defines which permissions are required by this Identity Provider to work
+     *
+     * @return the required permissions
+     */
+    public abstract String[] getRequiredPermissions();
+
+    /**
+     * Retry the last permission request issued to the user. If there is no recent
+     * request, this method will do nothing.
+     */
+    public void retryLastPermissionRequest() {
```

Comment:
It's a helper method for the user to retry the last request:
This is to be used when a request fails because the user canceled or denied it, because he doesn't know why we are needing that android permission. The person who implements the new `AuthorizedIdentityProvider` must implement the `onPermissionRequireExplanation()` method in order to let the final user know the reason of the request. In that method, you would normally explain the reason and give the final user a button to retry the request. If this is the case, then from this method you don't have the `activity` context neither the `serviceName` to perfom a call to `start()`. This is the reason why this helper method exists.